### ISSUE TITLE
  error: duplicate lang item in crate `sp_io` (which `sp_application_crypto` depends on)

### ISSUE THREAD

**Issue by aman09v** [paritytech/subport#150]
```
  error: duplicate lang item in crate `sp_io` (which `sp_application_crypto` depends on): `panic_impl`.
    |
    = note: the lang item is first defined in crate `std` (which `runtime_api` depends on)
    = note: first definition in `std` loaded from /root/.rustup/toolchains/nightly-x86_64-unknown-linux-gnu/lib/rustlib/wasm32-unknown-unknown/lib/libstd-c154c491353fefeb.rlib
    = note: second definition in `sp_io` loaded from /home/knoldus/sonata/target/release/wbuild/node-template-runtime/target/wasm32-unknown-unknown/release/deps/libsp_io-e846ce320076dcdd.rmeta


```

I get the above error when i compile my substrate project.
Can you tell what could be the problem?

**Comment by Veniamin** [comment]
Check std feature, this error is mostly about that some dependencies are missed in std feature in cargo.toml

**Comment by NukeManDan** [comment]
@aman09v if you are still having issues, please let us know. I am closing due to inactivity.

**Comment by Shubham9517** [comment]
@NukeManDan  I have the similar error and have almost applied every possible solution available on stack and github. I do have different version of dependencies in cargo.lock but I don't know how to remove it. I have already synced all the dependencies in cargo.toml to same version but still getting the same error.

This is the error that I am getting: 
error: duplicate lang item in crate `sp_io` (which `bp_runtime` depends on): `panic_impl`.
    |
    = note: the lang item is first defined in crate `sp_io` (which `frame_support` depends on)
    = note: first definition in `sp_io` loaded from /home/user/edition2021/2/substrate/target/release/wbuild/node-runtime/target/wasm32-unknown-unknown/release/deps/libsp_io-5e205f0a11762220.rmeta
    = note: second definition in `sp_io` loaded from /home/user/edition2021/2/substrate/target/release/wbuild/node-runtime/target/wasm32-unknown-unknown/release/deps/libsp_io-dfc24a7f2860dde8.rmeta

  error: duplicate lang item in crate `sp_io` (which `bp_runtime` depends on): `oom`.
    |
    = note: the lang item is first defined in crate `sp_io` (which `frame_support` depends on)
    = note: first definition in `sp_io` loaded from /home/user/edition2021/2/substrate/target/release/wbuild/node-runtime/target/wasm32-unknown-unknown/release/deps/libsp_io-5e205f0a11762220.rmeta
    = note: second definition in `sp_io` loaded from /home/user/edition2021/2/substrate/target/release/wbuild/node-runtime/target/wasm32-unknown-unknown/release/deps/libsp_io-dfc24a7f2860dde8.rmeta

**Comment by Shubham9517** [comment]
pls reply

**Comment by NukeManDan** [comment]
It's leaking in from upstream, have you checked that all things you depend on that subsequently depend on substrate also use the same dependency? This is discoverable with your lock file for what offending version is being used by what crate.

**Comment by Shubham9517** [comment]
I have checked all of the cargo.toml files in my project, the sp-io is used in 90 places and all consist of this dependency.
sp-io = { version = "4.0.0-dev", default-features = false, path = "../../../primitives/io" }

And also can you clear what do you mean by upstream and the packages that are conflicting in cargo.lock are:
(After cargo build I am able to see these 2 sp-io packages which is a kind of ambiguous)
[[package]]
name = "sp-io"
version = "4.0.0-dev"
dependencies = [
 "futures 0.3.17",
 "hash-db",
 "libsecp256k1 0.6.0",
 "log 0.4.14",
 "parity-scale-codec",
 "parking_lot 0.11.2",
 "sp-core 4.0.0-dev",
 "sp-externalities 0.10.0-dev",
 "sp-keystore 0.10.0-dev",
 "sp-runtime-interface 4.0.0-dev",
 "sp-state-machine 0.10.0-dev",
 "sp-std 4.0.0-dev",
 "sp-tracing 4.0.0-dev",
 "sp-trie 4.0.0-dev",
 "sp-wasm-interface 4.0.0-dev",
 "tracing",
 "tracing-core",
]

[[package]]
name = "sp-io"
version = "4.0.0-dev"
source = "git+https://github.com/paritytech/substrate?branch=master#755569d202b4007179cc250279bad55df45b5f7d"
dependencies = [
 "futures 0.3.17",
 "hash-db",
 "libsecp256k1 0.6.0",
 "log 0.4.14",
 "parity-scale-codec",
 "parking_lot 0.11.2",
 "sp-core 4.0.0-dev (git+https://github.com/paritytech/substrate?branch=master)",
 "sp-externalities 0.10.0-dev (git+https://github.com/paritytech/substrate?branch=master)",
 "sp-keystore 0.10.0-dev (git+https://github.com/paritytech/substrate?branch=master)",
 "sp-runtime-interface 4.0.0-dev (git+https://github.com/paritytech/substrate?branch=master)",
 "sp-state-machine 0.10.0-dev (git+https://github.com/paritytech/substrate?branch=master)",
 "sp-std 4.0.0-dev (git+https://github.com/paritytech/substrate?branch=master)",
 "sp-tracing 4.0.0-dev (git+https://github.com/paritytech/substrate?branch=master)",
 "sp-trie 4.0.0-dev (git+https://github.com/paritytech/substrate?branch=master)",
 "sp-wasm-interface 4.0.0-dev (git+https://github.com/paritytech/substrate?branch=master)",
 "tracing",
 "tracing-core",
]

**Comment by asmitadhungana** [comment]
Hi
I'm facing the same error while following the tutorial:
[Build a proof of existence dApp](https://docs.substrate.io/tutorials/v3/proof-of-existence/)

The error I'm getting is: 

` error: duplicate lang item in crate `sp_io` (which `frame_support` depends on): `panic_impl`.
    |
    = note: the lang item is first defined in crate `std` (which `pallet_template` depends on)
    = note: first definition in `std` loaded from /user/.rustup/toolchains/nightly-aarch64-apple-darwin/lib/rustlib/wasm32-unknown-unknown/lib/libstd-10bfc64443a2c69d.rlib
    = note: second definition in `sp_io` loaded from /user/work/learn/substrate/substrate-node-template/target/debug/wbuild/node-template-runtime/target/wasm32-unknown-unknown/release/deps/libsp_io-074a542bc0dbd659.rmeta

  error: duplicate lang item in crate `sp_io` (which `frame_support` depends on): `oom`.
    |
    = note: the lang item is first defined in crate `std` (which `pallet_template` depends on)
    = note: first definition in `std` loaded from /user/.rustup/toolchains/nightly-aarch64-apple-darwin/lib/rustlib/wasm32-unknown-unknown/lib/libstd-10bfc64443a2c69d.rlib
    = note: second definition in `sp_io` loaded from /user/work/learn/substrate/substrate-node-template/target/debug/wbuild/node-template-runtime/target/wasm32-unknown-unknown/release/deps/libsp_io-074a542bc0dbd659.rmeta

  error: could not compile `pallet-template` due to 2 previous errors
  warning: build failed, waiting for other jobs to finish...
  error: build failed`

**Comment by faizsarwar** [comment]
i am also facing same error

 error: duplicate lang item in crate `std` (which `serde` depends on): `panic_impl`.
    |
    = note: the lang item is first defined in crate `sp_io` (which `ocw_demo` depends on)
    = note: first definition in `sp_io` loaded from /home/faiz/rust/recipies book/recipes/target/release/wbuild/ocw-runtime/target/wasm32-unknown-unknown/release/deps/libsp_io-0713151cf511a047.rmeta
    = note: second definition in `std` loaded from /home/faiz/.rustup/toolchains/nightly-x86_64-unknown-linux-gnu/lib/rustlib/wasm32-unknown-unknown/lib/libstd-f812521814d28d7c.rlib

  error: duplicate lang item in crate `std` (which `serde` depends on): `oom`.
    |
    = note: the lang item is first defined in crate `sp_io` (which `ocw_demo` depends on)
    = note: first definition in `sp_io` loaded from /home/faiz/rust/recipies book/recipes/target/release/wbuild/ocw-runtime/target/wasm32-unknown-unknown/release/deps/libsp_io-0713151cf511a047.rmeta
    = note: second definition in `std` loaded from /home/faiz/.rustup/toolchains/nightly-x86_64-unknown-linux-gnu/lib/rustlib/wasm32-unknown-unknown/lib/libstd-f812521814d28d7c.rlib

  error: could not compile `ocw-demo` due to 2 previous errors
warning: build failed, waiting for other jobs to finish...
^C  Building [=======================> ] 888/908: librocksdb-sys(build), sc-network

**Comment by KavinduShamalka** [comment]
@faizsarwar Have you fix this error??

**Comment by sacha-l** [comment]
Have you tried doing `cargo clean` and then re-running the command to build your node, such as `cargo build --release`?

**Comment by tauruswei** [comment]
> i am also facing same error
> 
> error: duplicate lang item in crate `std` (which `serde` depends on): `panic_impl`. | = note: the lang item is first defined in crate `sp_io` (which `ocw_demo` depends on) = note: first definition in `sp_io` loaded from /home/faiz/rust/recipies book/recipes/target/release/wbuild/ocw-runtime/target/wasm32-unknown-unknown/release/deps/libsp_io-0713151cf511a047.rmeta = note: second definition in `std` loaded from /home/faiz/.rustup/toolchains/nightly-x86_64-unknown-linux-gnu/lib/rustlib/wasm32-unknown-unknown/lib/libstd-f812521814d28d7c.rlib
> 
> error: duplicate lang item in crate `std` (which `serde` depends on): `oom`. | = note: the lang item is first defined in crate `sp_io` (which `ocw_demo` depends on) = note: first definition in `sp_io` loaded from /home/faiz/rust/recipies book/recipes/target/release/wbuild/ocw-runtime/target/wasm32-unknown-unknown/release/deps/libsp_io-0713151cf511a047.rmeta = note: second definition in `std` loaded from /home/faiz/.rustup/toolchains/nightly-x86_64-unknown-linux-gnu/lib/rustlib/wasm32-unknown-unknown/lib/libstd-f812521814d28d7c.rlib
> 
> error: could not compile `ocw-demo` due to 2 previous errors warning: build failed, waiting for other jobs to finish... ^C Building [=======================> ] 888/908: librocksdb-sys(build), sc-network

Have you fix the problem？？

**Comment by AuroraLantean** [comment]
make sure all dependencies support `no_std`

**Comment by sgerodes** [comment]
I encountered the same issue and managed to resolve it. The problem arises because Cargo treats dependencies as different if they are specified with different Git references—even if they point to the same repository and code. This means that using a mix of tags, branches, or commit SHAs can cause Cargo to pull in multiple instances of the same crate, leading to conflicts like the duplicate lang item error. I had two instances of sp-io even though both point to the stable2407 version of the sdk.

I fixed the problem by ensuring all my dependencies consistently reference the same Git branch. Specifically, I replaced all instances of dependencies pointing to a tag or specific commit with references to the branch. Here's what I did:

For example, I replaced:
`sp-io = { git = "https://github.com/paritytech/polkadot-sdk.git", tag = "polkadot-stable2407", default-features = false }`
with:
`sp-io = { git = "https://github.com/paritytech/polkadot-sdk.git", branch = "stable2407", default-features = false}`

Did thatt with all other dependecies. 

Then I removed the lock and all build artifacts. Andf rebuilt the project.
```
rm Cargo.lock
cargo clean
cargo build --release
```
